jscomp/others/js_json_decode.mli

```diff
@@ -0,0 +1,291 @@
+  open Js.Json
+  let _ =
+
+  (* prints [Ok 1.23] *)
+  Js.log \@\@ parse "1.23" |> Decode.float
+
+  (* prints [Ok 23.] *)
+  Js.log \@\@ parse "23" |> Decode.float
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "true" |> Decode.float
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "null" |> Decode.float
+]}
+*)
+
+val int : int decoder
+(** Decodes a JSON value into an [int]
+    
+{b Returns} [Ok of int] if the JSON value is an integer, [Error of string] otherwise.
+
+@example {[
+  open Js.Json
+  let _ =
+
+  (* prints [Ok 23] *)
+  Js.log \@\@ parse "23" |> Decode.int
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "1.23" |> Decode.int
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "true" |> Decode.int
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "null" |> Decode.int
+]}
+*)
+
+val string : string decoder
+(** Decodes a JSON value into a [string]
+    
+{b Returns} [Ok of string] if the JSON value is a string, [Error of string] otherwise.
+
+@example {[
+  open Js.Json
+  let _ =
+
+  (* prints [Ok "foo"] *)
+  Js.log \@\@ parse "\"foo\"" |> Decode.string
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "1.23" |> Decode.string
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "null" |> Decode.string
+]}
+*)
+
+val nullable : 'a decoder -> 'a Js.null decoder
+(** Decodes a JSON value into an ['a Js.null]
+    
+{b Returns} [Ok Js.null] if the JSON value is [null], [Ok of 'a Js.null] if the
+given decoder succeeds, [Error of string] otherwise.
+
+@example {[
+  open Js.Json
+  let _ =
+
+  (* prints [Ok (Js.Null.return 23)] *)
+  Js.log \@\@ parse "23" |> Decode.(nullable int)
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "1.23" |> Decode.(nullable int)
+
+  (* prints [Ok Js.null] *)
+  Js.log \@\@ parse "null" |> Decode.(nullable int)
+]}
+*)
+
+val nullAs : 'a -> 'a decoder
+(** Returns the given value if the JSON value is [null]
+    
+{b Returns} [Ok of 'a] if the JSON value is [null], [Error of string] otherwise.
+
+@example {[
+  open Js.Json
+  let _ =
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "\"x\"" |> Decode.nullAs "x"
+
+  (* prints [Ok "x"] *)
+  Js.log \@\@ parse "null" |> Decode.nullAs "x"
+
+  (* prints [Ok None] *)
+  Js.log \@\@ parse "null" |> Decode.nullAs None
+]}
+*)
+
+val array : 'a decoder -> 'a array decoder
+(** Decodes a JSON array into an ['a array] using the given decoder on each array element
+    
+{b Returns} [Ok of 'a array] if the JSON value is a JSON array and all its
+elements are successfully decoded, [Error of string] otherwise.
+
+@example {[
+  open Js.Json
+  let _ =
+
+  (* prints [Ok [| 1; 2; 3 |]] *)
+  Js.log \@\@ parse "[1, 2, 3]" |> Decode.(array int)
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "[1, 2, "c"]" |> Decode.(array int)
+
+  (* prints [Error ...] *)
+  Js.log \@\@ parse "123" |> Decode.(array int)
+
+  (* prints [Ok None] *)
+  Js.log \@\@ parse "null" |> Decode.(array int)
+]}
+*)
+
+val dict : 'a decoder -> 'a Js.Dict.t decoder
```

Comment:
I imagine some ppx/codegen based decoder can be way more efficient than such pattern, let's not rush to bring in such API